server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1935,7 +1947,9 @@ public void openEngineAndRecoverFromTranslog() throws IOException {
      */
     public void openEngineAndSkipTranslogRecovery() throws IOException {
         assert routingEntry().recoverySource().getType() == RecoverySource.Type.PEER : "not a peer recovery [" + routingEntry() + "]";
-        recoveryState.validateCurrentStage(RecoveryState.Stage.TRANSLOG);
+        recoveryState.validateCurrentStage(
```

Comment:
... I think I'd prefer to do this in `TRANSLOG` stage even for unpromotable shards.